### PR TITLE
add basic check of obj.common properties

### LINKED ISSUE #669: Make sure object common properties are correct type
We now had the first adapter that created common.name as number and this had some side effects because most code assumes correct type here. 

So we should:
A) make sure that datatypes are correct by converting them hidden to correct type OR
B) generating errors and do not accept the object creation/change when wrong

What’s better in the end?!

I would live to do B that devs see their faults ... backward compatibility wise it would by A ;-)

### PR DESCRIPTION
**Pull request by foxriver76**
- fixes #669
- from my point of view, these are the most important checks, if anything is missing let me know

### CODE REVIEW

File: lib/tools.js [hunk 1]
```diff
@@ -1932,6 +1932,49 @@ function removeIdFromAllEnums(objects, id) {
...
+
+    if (obj.common.type !== undefined) {
+        if (typeof obj.common.type !== 'string') {
+            throw new Error(`obj.common.type has wrong type ${typeof obj.common.type} but has to be string`);
+        }
+
+        // if object type indicates a state, check that common type matches
+        const allowedStateTypes = ['number', 'string', 'boolean', 'array', 'object', 'mixed', 'file'];
+        if (obj.type === 'state' && !allowedStateTypes.includes(obj.common.type)) {
+            throw new Error(`obj.common.type has an invalid value (${obj.common.type}) but has to be one of ${allowedStateTypes.toString()}`);
```
Comment (AlCalzone) [inline lib/tools.js]:
```diff
- allowedStateTypes.toString()
+ allowedStateTypes.join(', ')
```

File: lib/tools.js [hunk 2]
```diff
@@ -1932,6 +1932,49 @@ function removeIdFromAllEnums(objects, id) {
...
+ */
+function validateObjectCommonProperties(obj) {
+    if (!obj || !obj.common) return;
+
+    if (obj.common.name !== undefined && typeof obj.common.name !== 'string') {
+        throw new Error(`obj.common.name has wrong type (${typeof obj.common.name}) but has to be string`);
+    }
+
+    if (obj.common.type !== undefined) {
+        if (typeof obj.common.type !== 'string') {
```
Comment (AlCalzone) [inline lib/tools.js]: Shouldn't we also check that one of the allowed types was passed?
https://github.com/ioBroker/ioBroker.docs/blob/master/docs/en/dev/objectsschema.md#object-types

File: lib/tools.js [hunk 3]
```diff
@@ -1932,6 +1932,49 @@ function removeIdFromAllEnums(objects, id) {
...
+ * If attributes are not provided, no error is thrown
+ *
+ * @param {object} obj an object which will be validated
+ * @throws Error if a property has the wrong type
+ */
+function validateObjectCommonProperties(obj) {
+    if (!obj || !obj.common) return;
+
+    if (obj.common.name !== undefined && typeof obj.common.name !== 'string') {
+        throw new Error(`obj.common.name has wrong type (${typeof obj.common.name}) but has to be string`);
```
Comment (AlCalzone) [inline lib/tools.js]: Let's try to be consistent with our error messages. So like #704, I propose
```
obj.common.name has an invalid type! Expected "string", received "${typeof obj.common.name}"
```
Same goes for the other messages. Maybe after these two PRs are merged, we can even refactor those into a single method `throwWrongType(propertyName, receivedType, expectedType)`.

File: lib/setup/setupInstall.js [hunk 4]
```diff
@@ -1153,6 +1153,15 @@ function Install(options) {
                         function setObjs() {
                             if (objs.length > 0) {
                                 const obj = objs.pop();
+
+                                try {
+                                    tools.validateObjectCommonProperties(obj);
+                                } catch (e) {
+                                    // todo: in the future we will not create this object
+                                    console.warn(`Type check failed on ${obj._id}: ${e.message}`);
```
Comment (AlCalzone) [inline lib/setup/setupInstall.js]: I propose the following message:
```
Object ${obj._id} is invalid: ${e.message}
```

File: lib/tools.js (hunk 1 above)
Comment (AlCalzone) [reply lib/tools.js]: Just to make the intent clearer

File: lib/tools.js (hunk 2 above)
Comment (foxriver76) [reply lib/tools.js]: We can do so, I was focusing on the `common` properties.

Comment (AlCalzone) [reply lib/tools.js]: Oh, right... I confused that with `obj.type`. Then I guess its fine here.

But I would still add the check for obj.type, doesn't hurt.

**Comment by Apollon77** [comment]
@foxriver76 I have merged now all the other PRs and it seems to need to check a conflict ... sorry

File: lib/tools.js (hunk 2 above)
Comment (foxriver76) [reply lib/tools.js]: I made the function a bit more general and also checking the type now, due to the fact that type has to exist on every created object and has to be valid. There hasn't been a check until now.

File: lib/tools.js (hunk 1 above)
Comment (foxriver76) [reply lib/tools.js]: Good point

File: lib/tools.js [hunk 5]
```diff
@@ -1932,6 +1932,63 @@ function removeIdFromAllEnums(objects, id) {
...
+function validateGeneralObjectProperties(obj) {
+    if (!obj || obj.type === undefined) {
+        throw new Error(`obj.type has to exist`);
+    }
+
+    if (typeof obj.type !== 'string') {
+        throw new Error(`obj.type has an invalid type! Expected "string", received  "${typeof obj.type}"`);
+    }
+
+    const allowedObjectTypes = ['state', 'channel', 'device', 'enum', 'host', 'adapter', 'instance', 'meta', 'config', 'script', 'user', 'group', 'chart'];
```
Comment (Apollon77) [inline lib/tools.js]: type "folder" is missing ... it is kind of "if device and channel is noth (clear) enough and you need an additional level"

Comment (Apollon77) [reply lib/tools.js]: Maybe also missing in the docs?

**Comment by Apollon77** [comment]
@GermanBluefox can you please also check over the list of object types and state types if there is maybe something undocumented that we missed? ;)

File: lib/tools.js (hunk 5 above)
Comment (foxriver76) [reply lib/tools.js]: It seems so.. I took the information from https://github.com/ioBroker/ioBroker.docs/blob/master/docs/en/dev/objectsschema.md#object-types

Comment (AlCalzone) [reply lib/tools.js]: > if device and channel is noth (clear) enough and you need an additional level

I love undocumented features :imp: . I have that exact case in my tradfri adapter. "Solved" it by not using an object at all for the additional level.

Comment (Apollon77) [reply lib/tools.js]: *gg*

Comment (GermanBluefox) [reply lib/tools.js]: "Folder" type was added https://github.com/ioBroker/ioBroker.docs/blob/master/docs/en/dev/objectsschema.md#object-types

File: lib/tools.js [hunk 6]
```diff
@@ -1932,6 +1932,63 @@ function removeIdFromAllEnums(objects, id) {
...
+    if (!allowedObjectTypes.includes(obj.type)) {
+        throw new Error(`obj.type has an invalid value (${obj.type}) but has to be one of ${allowedObjectTypes.join(', ')}`);
+    }
+
+    // obj.common is optional in general check
+    if (!obj.common) return;
+
+    if (obj.common.name !== undefined && typeof obj.common.name !== 'string') {
```
Comment (GermanBluefox) [inline lib/tools.js]: Name can be an object with many languages: `{"en": "Lamp", "de": "Lampe"}`

File: lib/tools.js [hunk 7]
```diff
@@ -1932,6 +1932,63 @@ function removeIdFromAllEnums(objects, id) {
...
+        throw new Error(`obj.common.name has an invalid type! Expected "string", received  "${typeof obj.common.name}"`);
+    }
+
+    if (obj.common.type !== undefined) {
+        if (typeof obj.common.type !== 'string') {
+            throw new Error(`obj.common.type has an invalid type! Expected "string", received  "${typeof obj.common.type}"`);
+        }
+
+        // if object type indicates a state, check that common.type matches
+        const allowedStateTypes = ['number', 'string', 'boolean', 'array', 'object', 'mixed', 'file'];
```
Comment (GermanBluefox) [inline lib/tools.js]: I think "json" could be too.. Must check the "admin"

Comment (GermanBluefox) [reply lib/tools.js]: Ok. no "json" type found.

Comment (Apollon77) [reply lib/tools.js]: But I also remember from old SCHEMA.md that "jso"N was there as state type

Comment (foxriver76) [reply lib/tools.js]: Hm there was also no `json` when I replaced SCHEMA.md today https://github.com/ioBroker/ioBroker/commit/ff4ad136ea06019309de8b336fff12fb324f59ed#diff-67a5df360d29ac792619774dd70c287eL200

File: lib/tools.js (hunk 6 above)
Comment (foxriver76) [reply lib/tools.js]: true, if an object is provided, is attribute 'en' mandatory?

File: lib/tools.js (hunk 7 above)
Comment (foxriver76) [reply lib/tools.js]: @Apollon77 but what would be the difference between json and object?

Comment (Apollon77) [reply lib/tools.js]: good question :-) I think noone uses it ... so lets leave it out, and wen will see warnings

**Pull request merged**